### PR TITLE
eve/log: FTP logging

### PR DESCRIPTION
**Pull request by jlucovsky**
This changeset is for issue [2315](https://redmine.openinfosecfoundation.org/issues/2315)  and implements eve logging for FTP, including documentation updates.

Continuation of #3962

Updated to address review comments and incorporate contribution from @imzke
- Break multiline responses into separate reply array entries

FTP logging adds 2 event types:
- ftp
- ftp_data

Each ftp event can contain:
- command
- command_data
- reply (in array format to handle multi-line replies)
- completion_code
- dynamic_port
- mode
- reply_received

ftp_data events contain:

- filename
- command associated with the data transfer.

## Sample `event_type ftp` output
```
{
  "timestamp": "2018-07-08T14:48:22.895430-0400",
  "flow_id": 1667103843128241,
  "pcap_cnt": 2911,
  "event_type": "ftp",
  "src_ip": "2601:0191:8500:2e00:7c0f:78e0:dc5b:f7c1",
  "src_port": 36874,
  "dest_ip": "2a00:0800:1010:0000:0000:0000:0000:0001",
  "dest_port": 21,
  "proto": "TCP",
  "tx_id": 12,
  "ftp": {
    "command": "RETR",
    "command_data": "100KB.zip",
    "reply": [
      "Opening BINARY mode data connection for 100KB.zip (102400 bytes).",
      "Transfer complete."
    ],
    "completion_code": [
      "150",
      "226"
    ],
    "reply_received": "yes"
  }
}
{
  "timestamp": "2019-03-10T18:20:17.762805-0400",
  "flow_id": 655371433876566,
  "pcap_cnt": 54028,
  "event_type": "ftp",
  "src_ip": "2a01:0e34:ee97:b130:8c3e:45ea:5ac6:e301",
  "src_port": 38970,
  "dest_ip": "2001:0700:0100:0118:0000:0000:0000:0047",
  "dest_port": 21,
  "proto": "TCP",
  "tx_id": 20,
  "ftp": {
    "command": "EPRT",
    "command_data": "|2|2a01:e34:ee97:b130:8c3e:45ea:5ac6:e301|41813|",
    "reply": [
      "EPRT command successful. Consider using EPSV."
    ],
    "completion_code": [
      "200"
    ],
    "dynamic_port": 41813,
    "mode": "active",
    "reply_received": "yes"
  }
}
{
  "timestamp": "2019-03-10T18:20:18.000534-0400",
  "flow_id": 655371433876566,
  "pcap_cnt": 54047,
  "event_type": "ftp",
  "src_ip": "2a01:0e34:ee97:b130:8c3e:45ea:5ac6:e301",
  "src_port": 38970,
  "dest_ip": "2001:0700:0100:0118:0000:0000:0000:0047",
  "dest_port": 21,
  "proto": "TCP",
  "tx_id": 21,
  "ftp": {
    "command": "RETR",
    "command_data": "index.html",
    "reply": [
      "Opening BINARY mode data connection for index.html (6712 bytes).",
      "Transfer complete."
    ],
    "completion_code": [
      "150",
      "226"
    ],
    "reply_received": "yes"
  }
}
{
  "timestamp": "2015-12-29T20:18:38.136221-0500",
  "flow_id": 1134263029039802,
  "pcap_cnt": 968367,
  "event_type": "ftp",
  "src_ip": "192.168.122.130",
  "src_port": 49191,
  "dest_ip": "162.144.12.241",
  "dest_port": 21,
  "proto": "TCP",
  "tx_id": 5,
  "ftp": {
    "command": "STOR",
    "command_data": "article_qpwfex967297_1451438317.zip",
    "reply": [
      "Accepted data connection\r\n",
      "6139779 Kbytes used (11%) - authorized: 51200000 Kb\r\n",
      "File successfully transferred\r\n",
      "0.091 seconds (measured here), 16.61 Kbytes per second"
    ],
    "completion_code": [
      "150",
      "226",
      "226",
      "226"
    ],
    "reply_received": "no"
  }
}
{
  "timestamp": "2018-07-08T14:48:36.307016-0400",
  "flow_id": 1655329190286257,
  "pcap_cnt": 3074,
  "event_type": "ftp",
  "src_ip": "2601:0191:8500:2e00:7c0f:78e0:dc5b:f7c1",
  "src_port": 36874,
  "dest_ip": "2a00:0800:1010:0000:0000:0000:0000:0001",
  "dest_port": 21,
  "proto": "TCP",
  "tx_id": 17,
  "ftp": {
    "command": "QUIT",
    "reply": [
      "Goodbye."
    ],
    "completion_code": [
      "221"
    ],
    "reply_received": "yes"
  }
}
```

## Sample `event_type ftp_data` output
```
{
  "timestamp": "2019-03-10T15:20:01.500157-0700",
  "flow_id": 1615446949360413,
  "parent_id": 274790086810710,
  "pcap_cnt": 68,
  "event_type": "ftp_data",
  "src_ip": "2a01:0e34:ee97:b130:8c3e:45ea:5ac6:e301",
  "src_port": 53697,
  "dest_ip": "2001:0700:0100:0118:0000:0000:0000:0047",
  "dest_port": 51093,
  "proto": "TCP",
  "tx_id": 0,
  "ftp_data": {
    "filename": "uio-app.js",
    "command": "RETR"
  }
}
{
  "timestamp": "2019-03-10T15:20:18.000469-0700",
  "flow_id": 1343493916181127,
  "parent_id": 274790086810710,
  "pcap_cnt": 116,
  "event_type": "ftp_data",
  "src_ip": "2001:0700:0100:0118:0000:0000:0000:0047",
  "src_port": 20,
  "dest_ip": "2a01:0e34:ee97:b130:8c3e:45ea:5ac6:e301",
  "dest_port": 41813,
  "proto": "TCP",
  "tx_id": 0,
  "ftp_data": {
    "filename": "index.html",
    "command": "RETR"
  }
}
```

### CODE REVIEW

File: src/output-json-ftp.c [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+
+    AppLayerParserRegisterLogger(IPPROTO_TCP, ALPROTO_FTP);
+    AppLayerParserRegisterLogger(IPPROTO_TCP, ALPROTO_FTPDATA);
+
+    result.ctx = output_ctx;
+    result.ok = true;
+    return result;
+}
+
+#define OUTPUT_BUFFER_SIZE 65535
```
Comment (catenacyber) [inline src/output-json-ftp.c]: Should we rather reuse the same macro from output-json.c by placing it in output-json.h (and maybe renaming int `JSON_OUTPUT_BUFFER_SIZE` ?)

File: src/app-layer-ftp.h [hunk 2]
```diff
@@ -79,9 +80,21 @@ typedef enum {
...
+    /* must be last */
+    FTP_COMMAND_MAX
     /** \todo more if missing.. */
 } FtpRequestCommand;
+
+typedef struct FtpCommand_ {
+    FtpRequestCommand command;
+    const char *command_name_upper;
+    const char *command_name_lower;
+    const uint32_t command_length;
```
Comment (catenacyber) [inline src/app-layer-ftp.h]: `uint8_t` is enough here

File: src/output-json-ftp.c [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+    }
+
+    if (!TAILQ_EMPTY(&tx->response_list)) {
+        FTPString *response;
+        TAILQ_FOREACH(response, &tx->response_list, next) {
+            /* handle multiple lines within the response, \r\n delimited */
+            uint8_t *where = response->str;
+            uint16_t length = response->len;
+            uint16_t pos;
+            while ((pos = JsonGetNextLineFromBuffer((const char *)where, length)) != UINT16_MAX) {
```
Comment (catenacyber) [inline src/output-json-ftp.c]: Should this parsing rather be done in app-layer-ftp.c ?

File: src/app-layer-ftp.c [hunk 4]
```diff
@@ -938,6 +1228,7 @@ static DetectEngineState *FTPDataGetTxDetectState(void *vtx)
 {
     FtpDataState *ftp_state = (FtpDataState *)vtx;
     return ftp_state->de_state;
+
```
Comment (catenacyber) [inline src/app-layer-ftp.c]: nit ?

File: src/app-layer-ftp.c [hunk 5]
```diff
@@ -312,49 +421,32 @@ static int FTPGetLine(FtpState *state)
...
-
-    if (input_len > 5 && SCMemcmpLowercase("retr", input, 4) == 0) {
-        fstate->command = FTP_COMMAND_RETR;
-    }
-
-    if (input_len >= 4 && SCMemcmpLowercase("epsv", input, 4) == 0) {
-        fstate->command = FTP_COMMAND_EPSV;
-    }
+    for (int i = 0; i < FTP_COMMAND_MAX; i++) {
+        if (!FtpCommands[i].command_length) {
```
Comment (catenacyber) [inline src/app-layer-ftp.c]: Can we skip this test by using `FTP_COMMAND_MAX-1`?

File: src/app-layer-ftp.c [hunk 6]
```diff
@@ -312,49 +421,32 @@ static int FTPGetLine(FtpState *state)
...
+            break;
+        }
+        if (input_len >= FtpCommands[i].command_length &&
```
Comment (catenacyber) [inline src/app-layer-ftp.c]: Is not there some multi pattern matching algorithm already implemented in Suricata ? (using a dictionary structure)
This comes at the cost of some precomputation and memory but it saves many `SCMemcmpLowercase` at runtime

File: src/app-layer-ftp.c [hunk 7]
```diff
@@ -582,9 +712,26 @@ static int FTPParsePassiveResponseV6(Flow *f, FtpState *state, uint8_t *input, u
...
+ *
+ * 1yz   Positive Preliminary reply
+ *
+ *                The requested action is being initiated; expect another
+ *                               reply before proceeding with a new command
+ */
+static inline bool FTPIsPPR(uint8_t *input, uint32_t input_len)
+{
+    return input_len >= 4 && isdigit(input[0]) && input[0] == '1' &&
+           isdigit(input[1]) && isdigit(input[2]) &&isspace(input[3]);
```
Comment (catenacyber) [inline src/app-layer-ftp.c]: nit space missing `&&isspace(input[3])`

**Comment by catenacyber** [comment]
Are there some new suricata-verify tests for this ?

File: src/output-json-ftp.c (hunk 3 above)
Comment (jlucovsky) [reply src/output-json-ftp.c]: IMO, the parsing is best deferred until needed. If FTP logging is not enabled, then the parsing is not needed.

File: src/app-layer-ftp.h (hunk 2 above)
Comment (jlucovsky) [reply src/app-layer-ftp.h]: Good point.

File: src/app-layer-ftp.c (hunk 5 above)
Comment (jlucovsky) [reply src/app-layer-ftp.c]: Good point.

File: src/app-layer-ftp.c (hunk 7 above)
Comment (jlucovsky) [reply src/app-layer-ftp.c]: Good catch.

File: src/app-layer-ftp.c (hunk 6 above)
Comment (jlucovsky) [reply src/app-layer-ftp.c]: Good question -- @victorjulien is there?

File: src/app-layer-ftp.c (hunk 4 above)
Comment (jlucovsky) [reply src/app-layer-ftp.c]: Will address.

File: src/output-json-ftp.c (hunk 1 above)
Comment (jlucovsky) [reply src/output-json-ftp.c]: Good suggestion; I'll take a look at the details and see if it will work out ok.

File: src/app-layer-ftp.c (hunk 6 above)
Comment (victorjulien) [reply src/app-layer-ftp.c]: Yes, have a look at how the smtp parser takes the string command and uses mpm to get its id.

File: src/output-json-ftp.c (hunk 1 above)
Comment (jlucovsky) [reply src/output-json-ftp.c]: I'll make this change to each of the output modules handling JSON.

File: src/app-layer-ftp.c (hunk 6 above)
Comment (jlucovsky) [reply src/app-layer-ftp.c]: Issue 3077 created for command parsing performance improvement.

**Comment by jlucovsky** [comment]
Continued in #4007

File: src/output-json-ftp.c (hunk 3 above)
Comment (catenacyber) [reply src/output-json-ftp.c]: Ok. but then could output-json-ftp.c call some new parsing function in app-layer-ftp.c ?
This way, we will defer parsing until needed and the parsing will be in app-*.c files

**Pull request closed without merging**